Adding support for Microchip PIC32CX SG devices

MyGh64605 opened the pull request:
zephyr: main: microchip: pic32cxsg: Addition of Microchip device family 'PIC32CX_SG'

The Microchip PIC32CX_SG family of devices are register and pin compatible with the Atmel SAME54 family.
Therefore the PIC32CX_SG drivers are associated with the Atmel SAME54 drivers.
No new additional drivers have been added.

Tested build with PIC32CXSG41 and PIC32CXGS61 boards.
Processed PIC32CX_SG files utilizing command "find * -type f -print0 | xargs -0 dos2unix -k". 
Retested PIC32CXSG41 and PIC32CZSG61 builds.

Associated pull request for the addition of the PIC32CX_SG devices to hal_microchip:
https://github.com/zephyrproject-rtos/hal_microchip/pull/21

Signed-off-by: Michael D Sherwood <michael.sherwood@microchip.com>

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
#if defined(CONFIG_SOC_PIC32CX1025SG41100)
#include <pic32cx1025sg41100.h>
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
compatible = "same54,xpro", "atmel,same54p20a", "atmel,same54";
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
#include "../common/component/usb_component_fixup_pic32cxsg.h"
#include "../common/component/usb_component_fixup_pic32cxsg.h"
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Adding support for Microchip PIC32CX SG devices

MyGh64605 committed:
Use main updates

All changed to reflect updates from 3.6 to main

MyGh64605 committed:
Use main updates

All files changed to reflect the updates from main

MyGh64605 committed:
Tested build with PIC32CXSG41 and PIC32CXGS61 boards

Signed-off-by: Michael D Sherwood <michael.sherwood@microchip.com>

MyGh64605 committed:
Renamed 'mplabx' folder to 'pic32c'. Tested build with PIC32CXSG41 and PIC32CXGS61 boards

MyGh64605 committed:
Processed new PIC32CX_SG files utilizing command "find * -type f -print0 | xargs -0 dos2unix -k". Retested PIC32CXSG41 and PIC32CZSG61 builds

Signed-off-by: Michael D Sherwood <michael.sherwood@microchip.com>

MyGh64605 committed:
Merge branch 'main' of https://github.com/MyGh64605/zephyr

pmcgaugh committed:
Adding support for Microchip PIC32CX SG devices

Tested build with PIC32CXSG41 and PIC32CXGS61 boards
Processed new PIC32CX_SG files utilizing command "find * -type f -print0 | xargs -0 dos2unix -k". Retested PIC32CXSG41 and PIC32CZSG61 builds

Signed-off-by: Michael D Sherwood <michael.sherwood@microchip.com>

henrikbrixandersen reviewed (changes requested):
Please split this PR into an initial PR with basic SoC/board support (basically just what is needed for booting, logging and GPIOs) and leave the rest for smaller, incremental PRs.

pmcgaugh committed:
soc: pic32c: Remove undefined _UINT32_ type

_UINT32_ is undefined and causing compilation issues for the atmel same54 chip.

Signed-off-by: Peter McGaughey <peter.mcgaughey@daikincomfort.com>

nandojve commented on .gitignore:
```diff
@@ -91,3 +91,5 @@ ModulesMaintainers.txt
 Nits.txt
 Pylint.txt
 YAMLLint.txt
+/dts/arm/microchip/pic32cxsg
+pic32cx_sg41_cult
```
Hi @pmcgaugh ,

It is not clear the use case here. If you need some files to autogenerate other maybe a regular expression could be better to cover all the cases.

MyGh64605 replied:
@pmcgaugh @nandojve I did not add these lines, what do you want me to do ?
A meeting would be a option.

nandojve replied:
We can schedule a meeting to sync about. I can try after 18h ( UTC+2 ).

nandojve commented on boards/microchip/atmarktechno/degu_evk/Kconfig:
```diff
@@ -0,0 +1,10 @@
+# Degu Evaluation Kit configuration
```
This board should not be part of the PR. It seems to be from another manufacturer and it is out of scope.

MyGh64605 replied:
I did not add or modify these entries. What do you want me to do ?
As you said, this is out of scope, so I should not touch them.

nandojve replied:
You are submitting the code so you need Signed-off-by all commits and you will be responsible to make all review adjusts then. in this case, you must drop all this atmarktechno board.

nandojve commented on boards/microchip/pic32cxsg41_cult/support/openocd.cfg:
```diff
@@ -0,0 +1,25 @@
+$_TARGETNAME configure -event gdb-detach {
+	echo "Debugger detaching: resuming execution"
+	resume
+}
```
Missing new line

MyGh64605 replied:
Thanks for pointing this out. Resolved.

nandojve commented on drivers/adc/CMakeLists.txt:
```diff
@@ -16,10 +16,11 @@ zephyr_library_sources_ifdef(CONFIG_ADC_VF610		adc_vf610.c)
 zephyr_library_sources_ifdef(CONFIG_ADC_NRFX_ADC	adc_nrfx_adc.c)
 zephyr_library_sources_ifdef(CONFIG_ADC_NRFX_SAADC	adc_nrfx_saadc.c)
-zephyr_library_sources_ifdef(CONFIG_ADC_SAM		adc_sam.c)
+zephyr_library_sources_ifdef(CONFIG_ADC_SAM			adc_sam.c)
```
Is your editor using TAB 4 instead 8? Check all files to use correct indentation.

MyGh64605 replied:
No, editor is set to TAB 4

nandojve replied:
You should configure your IDE to use TAB 8 (zephyr default) and adjust if appropriated the indentation.

nandojve commented on drivers/adc/Kconfig.pic32cxsg:
```diff
@@ -0,0 +1,10 @@
+	depends on DT_HAS_PIC32CXSG_ADC_ENABLED
+	select ADC_CONFIGURABLE_INPUTS
+	help
+	  Enable Microchip PIC32CXSG MCU Family Analog-to-Digital Converter (ADC) driver.
```
new line

MyGh64605 replied:
Thanks for pointing this out. Resolved.

nandojve commented on drivers/can/Kconfig.pic32cxsg:
```diff
@@ -0,0 +1,9 @@
+	bool "Microchip PIC32CXSG CAN driver"
+	default y
+	depends on DT_HAS_PIC32CXSG_CAN_ENABLED
+	select CAN_MCAN
```
revise all files

nandojve commented on drivers/dac/dac_esp32.c:
```diff
@@ -64,7 +64,7 @@ static int dac_esp32_init(const struct device *dev)
 
 	if (clock_control_on(cfg->clock_dev,
-		(clock_control_subsys_t) &cfg->clock_subsys) != 0) {
+		(clock_control_subsys_t) cfg->clock_subsys) != 0) {
```
The change in this specific file should be on another PR. Drop it.

nandojve commented on drivers/dma/Kconfig.pic32cxsg:
```diff
@@ -0,0 +1,11 @@
+# Microchip PIC32CXSG DMAC configuration options
+
+# Copyright (c) 2018 Google LLC.
```
MCHP

nandojve commented on drivers/ethernet/Kconfig.pic32cxsg_gmac:
```diff
@@ -0,0 +1,105 @@
+	bool "Microchip PIC32CXSG Ethernet driver"
+	default y
+	depends on NET_BUF_FIXED_DATA_SIZE
+	depends on DT_HAS_MICROCHIP_PIC32CXSG_GMAC_ENABLED || \
+		   DT_HAS_MICROCHIP_PIC32CXSG_GMAC_ENABLED
```
2 times same chip?

nandojve commented on include/zephyr/drivers/can/can_mcan.h:
```diff
@@ -1297,7 +1297,7 @@ struct can_mcan_config {
 		.mram_offsets = CAN_MCAN_DT_MRAM_OFFSETS_GET(node_id),                             \
 		.mram_size = CAN_MCAN_DT_MRAM_ELEMENTS_SIZE(node_id),                              \
-		.custom = _custom,                                                                 \
+<		.custom = _custom,                                                                 \
```
typo?

nandojve commented on soc/atmel/sam0/common/atmel_sam0_dt.h:
```diff
@@ -1,6 +1,8 @@
 /*
  * Copyright (c) 2020 Linaro Ltd.
  * Copyright (c) 2021 Gerson Fernando Budke
+ * Copyright (c) 2024 Microchip added 'MCLK_APBAMASK_EIC_BIT_MASK'
+ * Copyright (c) 2024 Microchip added 'GCLK_PCHCTRL_CHEN_BIT_MASK'
```
```diff
- * Copyright (c) 2024 Microchip added 'MCLK_APBAMASK_EIC_BIT_MASK'
- * Copyright (c) 2024 Microchip added 'GCLK_PCHCTRL_CHEN_BIT_MASK'
+ * Copyright (c) 2024 Microchip
```

nandojve commented on soc/microchip/pic32cxsg/common/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+
+zephyr_sources_ifdef(CONFIG_BOOTLOADER_BOSSA bossa.c)
+
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_PIC32CXSG41 soc_pic32cxsg.c)
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_PIC32CXSG60 soc_pic32cxsg.c)
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_PIC32CXSG61 soc_pic32cxsg.c)
```
```diff
-zephyr_sources_ifdef(CONFIG_SOC_SERIES_PIC32CXSG41 soc_pic32cxsg.c)
-zephyr_sources_ifdef(CONFIG_SOC_SERIES_PIC32CXSG60 soc_pic32cxsg.c)
-zephyr_sources_ifdef(CONFIG_SOC_SERIES_PIC32CXSG61 soc_pic32cxsg.c)
+zephyr_sources_ifdef(CONFIG_SOC_FAMILY_MICROCHIP_PIC32CXSG soc_pic32cxsg.c)
```

nandojve commented on soc/microchip/pic32cxsg/common/Kconfig.pic32cxsg:
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) 2024 Microchip
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_SERIES_PIC32CXSG41 || SOC_SERIES_PIC32CXSG60 || SOC_SERIES_PIC32CXSG61
```
If the whole family covers then it is better to use CONFIG_SOC_FAMILY_MICROCHIP_PIC32CXSG.

MyGh64605 replied:
Thanks for pointing this out. Resolved.

nandojve commented on soc/microchip/pic32cxsg/common/bossa.c:
```diff
@@ -0,0 +1,61 @@
+/*
```
Ok for now!

nandojve commented on soc/microchip/pic32cxsg/common/component/ac_component_fixup_pic32cxsg.h:
```diff
@@ -0,0 +1,279 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _MICROCHIP_PIC32CXSG_AC_COMPONENT_FIXUP_H_
```
Shouldn't this be at hal?

MyGh64605 replied:
I am just following the way that sam0/common was used.

nandojve replied:
This is something that we should avoid always. We need to create specific commits on hal to make the correct adjustments. The commits should be component by component.

nandojve commented on soc/microchip/pic32cxsg/common/instance/ac_instance_fixup_pic32cxsg.h:
```diff
@@ -0,0 +1,47 @@
+#define _MICROCHIP_PIC32CXSG_AC_INSTANCE_FIXUP_H_
+
+/* ========== Register definition for AC peripheral ========== */
+#if (defined(__ASSEMBLY__) || defined(__IAR_SYSTEMS_ASM__))
```
Shouldn't this be at hal?

nandojve reviewed (changes requested):
Hi @MyGh64605,

I point out the minor things. I see a mix of SAM/PIC32 defintions which is not clear to me yet. If you want to introduce the bindings for the drivers using the microchip prefix you should use it. Otherwise you need to drop all unused things.

Since this can be a big change, I suggest to you drop at moment to make the PR small and easy to review.

If you prefer, we can schedule a call this week to sync about above.

BTW, I'm assuming that registers and devicetree address are correct.

MyGh64605 committed:
Merge branch 'main' of https://github.com/MyGh64605/zephyr

# Conflicts:
#	.github/workflows/bsim-tests.yaml
#	modules/Kconfig.microchip
#	soc/atmel/sam0/common/atmel_sam0_dt.h
#	west.yml

henrikbrixandersen commented:
> Please split this PR into an initial PR with basic SoC/board support (basically just what is needed for booting, logging and GPIOs) and leave the rest for smaller, incremental PRs. 

@MyGh64605 Before going any further, you _really_ need to split this into multiple PRs. 

First PR should only add initial support (basically hello_world and blinky support). Following up on this, you can add support by the various peripherals one by one.  

The rationale for this is a) smaller PRs are easier to review b) different subsystems require reviews from different maintainers. 

Please see https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#contributor-expectations

MyGh64605 committed:
Signed-off-by: Michael D Sherwood <michael.sherwood@microchip.com>
Update for 'final' pull request

nordicjm reviewed (changes requested):
Something very wrong with this PR, likely stray files that should not have been added

MyGh64605 committed:
Added svd files for J-LINK support

Signed-off-by: Michael D Sherwood <michael.sherwood@microchip.com>

nandojve commented:
Hi @MyGh64605 ,

To this PR get attention and move forward:

1- Follow the https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#contributor-expectations
2- The PR must correct associate https://github.com/zephyrproject-rtos/hal_microchip/pull/24. The 21 should be closed since 24 superseds.
3- The commits should have proper brief and description.
4- It is forbidden merge commits on your commit sequence. In this case, you need to rebase the code on top of mainline, solve all conflicts and have a clean commit history.

nandojve commented:
Hi @MyGh64605 ,

I was wondering if you have any update on this.

MyGh64605 commented:
@nandojve

If you want to close this PR, that is fine. This pull is so far behind it will be easier to start over.
We have just created a new repo with the latest zephyr content.
Will be making the additions to create the PIC32CXSG 'Hello World' projects. 
Then creating a new PR.

nandojve commented:
> @nandojve
> 
> If you want to close this PR, that is fine. This pull is so far behind it will be easier to start over. We have just created a new repo with the latest zephyr content. Will be making the additions to create the PIC32CXSG 'Hello World' projects. Then creating a new PR.

nandojve commented:
Hi @MyGh64605

> @nandojve
> 
> If you want to close this PR, that is fine. This pull is so far behind it will be easier to start over. We have just created a new repo with the latest zephyr content. Will be making the additions to create the PIC32CXSG 'Hello World' projects. Then creating a new PR.

Ok, I'll close this one as you mention.
